api/v1alpha3/webhook_test.go

```diff
@@ -0,0 +1,123 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"fmt"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api/util"
+)
+
+func TestClusterConversion(t *testing.T) {
+	g := NewWithT(t)
+	ns, err := testEnv.CreateNamespace(ctx, fmt.Sprintf("conversion-webhook-%s", util.RandomString(5)))
+	g.Expect(err).ToNot(HaveOccurred())
+	clusterName := fmt.Sprintf("test-cluster-%s", util.RandomString(5))
+	cluster := &Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: ns.Name,
+		},
+	}
+
+	g.Expect(testEnv.Create(ctx, cluster)).To(Succeed())
+	defer func(do ...client.Object) {
+		g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+	}(ns, cluster)
+}
+
+func TestMachineSetConversion(t *testing.T) {
+	g := NewWithT(t)
+	ns, err := testEnv.CreateNamespace(ctx, fmt.Sprintf("conversion-webhook-%s", util.RandomString(5)))
+	g.Expect(err).ToNot(HaveOccurred())
+
+	clusterName := fmt.Sprintf("test-cluster-%s", util.RandomString(5))
+	machineSetName := fmt.Sprintf("test-machineset-%s", util.RandomString(5))
+	machineSet := &MachineSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns.Name,
+			Name:      machineSetName,
+		},
+		Spec: MachineSetSpec{
+			ClusterName:     clusterName,
+			Template:        newDummyMachineTemplate(ns.Name, clusterName),
+			MinReadySeconds: 10,
+			Replicas:        pointer.Int32Ptr(1),
+			DeletePolicy:    "Random",
+		},
+	}
+
+	g.Expect(testEnv.Create(ctx, machineSet)).To(Succeed())
+	defer func(do ...client.Object) {
+		g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+	}(ns, machineSet)
+}
+func TestMachineDeploymentConversion(t *testing.T) {
+	g := NewWithT(t)
+	ns, err := testEnv.CreateNamespace(ctx, fmt.Sprintf("conversion-webhook-%s", util.RandomString(5)))
+	g.Expect(err).ToNot(HaveOccurred())
+
+	clusterName := fmt.Sprintf("test-cluster-%s", util.RandomString(5))
+	machineDeploymentName := fmt.Sprintf("test-machinedeployment-%s", util.RandomString(5))
+	machineDeployment := &MachineDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      machineDeploymentName,
+			Namespace: ns.Name,
+		},
+		Spec: MachineDeploymentSpec{
+			ClusterName: clusterName,
+			Template:    newDummyMachineTemplate(ns.Name, clusterName),
+			Replicas:    pointer.Int32Ptr(0),
+		},
+	}
+
+	g.Expect(testEnv.Create(ctx, machineDeployment)).To(Succeed())
+	defer func(do ...client.Object) {
+		g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+	}(ns, machineDeployment)
+}
+
+func newDummyMachineTemplate(namespace, clusterName string) MachineTemplateSpec {
```

Comment:
```suggestion
func newMachineTemplate(namespace, clusterName string) MachineTemplateSpec {
```
Eventually you can use Fake instead